go/cmd/exporter/worker.go

```diff
@@ -0,0 +1,246 @@
+package main
+
+import (
+	"archive/zip"
+	"bytes"
+	"cmp"
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"io"
+	"log/slog"
+	"maps"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/osv.dev/go/logger"
+	"github.com/ossf/osv-schema/bindings/go/osvschema"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	gitEcosystem        = "GIT"
+	vanirVulnsFilename  = "osv_git.json"
+	allZipFilename      = "all.zip"
+	modifiedCSVFilename = "modified_id.csv"
+	ecosystemsFilename  = "ecosystems.txt"
+)
+
+type ecosystemWorker struct {
+	ecosystem string
+	ch        chan *osvschema.Vulnerability
+}
+
+func newEcosystemWorker(ctx context.Context, ecosystem string, writeCh chan<- writeMsg, wg *sync.WaitGroup) *ecosystemWorker {
+	ch := make(chan *osvschema.Vulnerability)
+	worker := &ecosystemWorker{
+		ecosystem: ecosystem,
+		ch:        ch,
+	}
+	go worker.run(ctx, writeCh, wg)
+
+	return worker
+}
+
+type vulnData struct {
+	id   string
+	data []byte
+}
+
+var protoMarshaller = protojson.MarshalOptions{
+	UseProtoNames: true, // TODO(michaelkedar): https://github.com/ossf/osv-schema/pull/442
+}
+
+func (w *ecosystemWorker) run(ctx context.Context, writeCh chan<- writeMsg, wg *sync.WaitGroup) {
+	defer wg.Done()
+	logger.Info("new ecosystem worker started", slog.String("ecosystem", w.ecosystem))
+	var allVulns []vulnData
+	var csvData [][]string
+	var vanirVulns []vulnData
+	for {
+		var v *osvschema.Vulnerability
+		var ok bool
+
+		// Wait to receive a vulnerability, or be cancelled.
+		select {
+		case v, ok = <-w.ch:
```

Comment:
hmm feels like this should also just be a for loop over w.ch
Then all the final writeCSV and writeZIP functions can be put at the bottom of this function after the loop